### PR TITLE
Added UC8 fit, predict, vector to points

### PR DESCRIPTION
**Pull request by ValentinaHutter**

### CODE REVIEW

File: src/openeo_processes/cubes.py [hunk 1]
```diff
@@ -1929,4 +1933,126 @@ def exec_xar(data, process, dimension, target_dimension=None, context=None):
...
+@process
+def fit_random_forest():
+
+    return FitRandomForest()
+
+
+class FitRandomForest:
+
+    @staticmethod
+    def exec_xar(target, predictor, training = 50, rfparameters = [], client = None):
```
Comment (LukeWeidenwalker) [inline src/openeo_processes/cubes.py:1952]: `training` here is a percentage, I think it should be required to be between 0-1 to stick to the conventions of the ML community (see https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html). Allowing this to be an integer could suggest that this is the absolute number of samples in the test set.

File: src/openeo_processes/cubes.py [hunk 2]
```diff
@@ -1929,4 +1933,126 @@ def exec_xar(data, process, dimension, target_dimension=None, context=None):
...
+            YY = YY.drop(columns=['spatial_ref'])
+        if training > 1:
+            training = training/100
+        X_train, X_test, y_train, y_test = train_test_split(df, YY, test_size=training)
+
+        dtrain = xgb.dask.DaskDMatrix(client, X_train, y_train)
+        dtest = xgb.dask.DaskDMatrix(client, X_test, y_test)
+        output = xgb.dask.train(client, params, dtrain, num_boost_round=4, evals=[(dtrain, "train")])
+
+        return output
```
Comment (LukeWeidenwalker) [inline src/openeo_processes/cubes.py:1985]: output is currently a dict with: ```{'booster': <xgboost.core.Booster at 0x7f8110b58220>,
 'history': {'train': OrderedDict([('rmse',
                [0.287989, 0.287337, 0.286719, 0.286118])])}}```
We'll need to convert this xgboost output object into what is required for this process!
The requirement from the function description is:
- The model itself 
- The internal validation (Out-of-Bag-Error) 
- The validation result 

As far as I understand it, an Out-of-Bag-Error (internal validation) is computed during each bootstrap sample, do we return the mean of these? So the "validation result" would be a RMSE on the test_samples that we put to the side with the train_test_split?

File: src/openeo_processes/cubes.py [hunk 3]
```diff
@@ -1929,4 +1933,126 @@ def exec_xar(data, process, dimension, target_dimension=None, context=None):
...
+        stacked = predictors.stack(z=C)
+        I = stacked['z'].values
+        DF = stacked.to_dataset(dim='bands').to_dask_dataframe()
+        df = DF.drop(columns=['z'])
+        if 'spatial_ref' in df:
+            df = df.drop(columns=['spatial_ref'])
+        ypred = xgb.dask.predict(client, model, df)
+        yf = ypred.to_frame()
+        ds = xr.Dataset()
+        da = ds.from_dataframe(yf)
```
Comment (LukeWeidenwalker) [inline src/openeo_processes/cubes.py:2014]: I think the naming of these variables wants to be more descriptive - one, because it makes this code easier to read and two, because above we have `import dask.array as da`, so there's potential to mix up these variables here!

File: src/openeo_processes/cubes.py [hunk 4]
```diff
@@ -1929,4 +1933,126 @@ def exec_xar(data, process, dimension, target_dimension=None, context=None):
...
+        YY = DF2.drop(columns=['z'])
+        if 'spatial_ref' in YY:
+            YY = YY.drop(columns=['spatial_ref'])
+        if training > 1:
+            training = training/100
+        X_train, X_test, y_train, y_test = train_test_split(df, YY, test_size=training)
+
+        dtrain = xgb.dask.DaskDMatrix(client, X_train, y_train)
+        dtest = xgb.dask.DaskDMatrix(client, X_test, y_test)
+        output = xgb.dask.train(client, params, dtrain, num_boost_round=4, evals=[(dtrain, "train")])
```
Comment (LukeWeidenwalker) [inline src/openeo_processes/cubes.py:1983]: `num_boost_round` wants to be set to 1, otherwise we're doing gradient boosting! (see https://xgboost.readthedocs.io/en/stable/tutorials/rf.html)

File: src/openeo_processes/cubes.py [hunk 5]
```diff
@@ -1929,4 +1933,126 @@ def exec_xar(data, process, dimension, target_dimension=None, context=None):
...
+        stacked = target.stack(z=C)
+        DF = stacked.to_dataset(dim='bands').to_dask_dataframe()
+        df = DF.drop(columns=['z'])
+        if 'spatial_ref' in df:
+            df = df.drop(columns=['spatial_ref'])
+        C = list(predictor.dims)
+        if 'bands' in C:
+            C.remove('bands')
+        stacked2 = predictor.stack(z=C)
+        DF2 = stacked2.to_dataset(dim='bands').to_dask_dataframe()
```
Comment (LukeWeidenwalker) [inline src/openeo_processes/cubes.py:1973]: I think at this point it would be useful to check that the dimensions of target and predictors fits and raise an Exception otherwise!

File: src/openeo_processes/cubes.py [hunk 6]
```diff
@@ -1929,4 +1933,126 @@ def exec_xar(data, process, dimension, target_dimension=None, context=None):
...
+        YY = DF2.drop(columns=['z'])
+        if 'spatial_ref' in YY:
+            YY = YY.drop(columns=['spatial_ref'])
+        if training > 1:
+            training = training/100
+        X_train, X_test, y_train, y_test = train_test_split(df, YY, test_size=training)
+
+        dtrain = xgb.dask.DaskDMatrix(client, X_train, y_train)
```
Comment (LukeWeidenwalker) [inline src/openeo_processes/cubes.py:1981]: How are we dealing with potential nans? I've read in https://github.com/Open-EO/openeo-processes/pull/306/files that this is what the `array_interpolate_linear` process is for. I'm not certain it's a good idea to make the user take care of that!

File: src/openeo_processes/cubes.py (hunk 1 above)
Comment (ValentinaHutter) [reply src/openeo_processes/cubes.py:1952]: For now I added a little workaround some lines below: 
`if training > 1:`
`            training = training/100`
so at the moment both values between 0 and 1 and 0 and 100 work, but I will change the default and ask our colleagues about it :)

**Comment by clausmichele** [comment]
May I ask you why you choose xgboost over scikit-learn ? For scalability?

File: src/openeo_processes/cubes.py [hunk 7]
```diff
@@ -1929,4 +1937,267 @@ def exec_xar(data, process, dimension, target_dimension=None, context=None):
...
+        if 'bands' in P:
+            P.remove('bands')
+        if P == C:
+            check = np.array([])
+            for dim in P:
+                check = np.append(check, (target[dim].values == predictors[dim].values).all())
+            if (check == False).any():
+                raise Exception('DimensionMismatch')
+        stacked = target.stack(z=C)
+        stacked2 = predictors.stack(z=C)
```
Comment (LukeWeidenwalker) [inline src/openeo_processes/cubes.py:2024]: I think this wants to be `stacked2 = predictors.stack(z=C)
```suggestion
        stacked2 = predictors.stack(z=P)
```

File: src/openeo_processes/cubes.py [hunk 8]
```diff
@@ -1929,4 +1937,267 @@ def exec_xar(data, process, dimension, target_dimension=None, context=None):
...
+        C = list(target.dims)
+        if 'bands' in C:
+            C.remove('bands')
+        P = list(predictors.dims)
+        if 'bands' in P:
+            P.remove('bands')
+        if P == C:
+            check = np.array([])
+            for dim in P:
+                check = np.append(check, (target[dim].values == predictors[dim].values).all())
```
Comment (LukeWeidenwalker) [inline src/openeo_processes/cubes.py:2020]: Maybe a dumb question, but wouldn't calling `.values` here load these into memory?

File: src/openeo_processes/cubes.py [hunk 9]
```diff
@@ -1929,4 +1937,267 @@ def exec_xar(data, process, dimension, target_dimension=None, context=None):
...
+# PredictRandomForest process
+###############################################################################
+
+@process
+def predict_random_forest():
+
+    return PredictRandomForest()
+
+
+class PredictRandomForest:
```
Comment (LukeWeidenwalker) [inline src/openeo_processes/cubes.py:2090]: This should probably become `PredictRegrRandomForest` too

Comment (ValentinaHutter) [reply src/openeo_processes/cubes.py:2088]: I think it is still called PredictRandomForest as we could use this function to predict random forest regression and random forest classifier, which is why it would not be renamed :)

File: src/openeo_processes/cubes.py (hunk 8 above)
Comment (ValentinaHutter) [reply src/openeo_processes/cubes.py:2029]: It is okay to load the values of the dimension labels, with target['time'].values for example all the names of temporal coordinates would be loaded, so this is fine. but we could not call target.values, which would load values into memory.

Comment (ValentinaHutter) [reply src/openeo_processes/cubes.py:2029]: It is okay to load the values of the dimension labels, with target['time'].values for example all the names of temporal coordinates would be loaded, so this is fine. but we could not call target.values, which would load values into memory.

Comment (ValentinaHutter) [reply src/openeo_processes/cubes.py:2029]: It is okay to load the values of the dimension labels, with target['time'].values for example all the names of temporal coordinates would be loaded, so this is fine. but we could not call target.values, which would load values into memory.

File: src/openeo_processes/cubes.py [hunk 10]
```diff
@@ -1931,4 +1939,293 @@ def exec_xar(data, process, dimension, target_dimension=None, context=None):
...
+@process
+def fit_regr_random_forest():
+
+    return FitRegrRandomForest()
+
+
+class FitRegrRandomForest:
+
+    @staticmethod
+    def exec_xar(predictors, target, training, num_trees = 100, mtry = None, client = None):
```
[the remark this answers is not in the document]
Comment (LukeWeidenwalker) [reply src/openeo_processes/cubes.py:2023]: Also if mtry is None, according to the process definition, we should set this to 
`"description": "Specifies how many split variables will be used at a node. Default value is `null`, which corresponds to the number of predictors divided by 3."`

[the remark this answers is not in the document]
Comment (LukeWeidenwalker) [reply src/openeo_processes/cubes.py:2023]: Also, mtry will come in as an integer specifying the number of columns, not a fraction like `colsample_bynode` expects, so we'll have to convert that accordingly!
